### ISSUE TITLE
Feature influence production rates of weapons

### ISSUE THREAD

**Issue by codingberlin** [jsettlers/settlers-remake#210]
I would like to implement the GUI and functionality to influence the production rates of weapons. (Because the computerplayer needs to influence it ;-) )

I would implement both: influence the ratios by increasing or decreasing the blue bars (Do we have somewhere such a bar already?) and production by prio (by increasing the numbers left to the blue bar)

I would focus on weapons but could implement tools as well, if we decide how to handle them in combination with the autoproduction of the toolsmith when tools are missing.
-> Should the autoproduction increase the numbers left to the blue bars so that the player is able to get feedback (and maybe influence it) what the toolsmith is producing?

**Comment by codingberlin** [comment]
@michaelzangl while working on that issue I got two questions. Maybe you can help me?

1) how can an action be fired by BarFill when it is clicked? (e.g. like with ```new Button(Action action, ...)```)
2) how can I implement that the ToolsPanel refreshes the displayed information? (e.g. you add 10 bows and 2 were produced the panel shows 10 until I move a bit, than it refreshes)

Tanks in advanced!

**Comment by michaelzangl** [comment]
1) Extend the BarFill. Override getAction(x, y) to construct a new action. Pass null as action to the constructor.
2) Use this class to help you (see usage in buildings menu for an example).
https://github.com/jsettlers/settlers-remake/blob/master/jsettlers.graphics/src/jsettlers/graphics/utils/UIUpdater.java

**Comment by codingberlin** [comment]
Both helped me. Thx!

By the way I coupled now the automatical and manual Toolproduction.
If tools for workers are missing, they will be produced in their priority order. But now they were displayed in the ToolsPanel as well. If the player reduces them, they will be increased again (because the tool is missing). If no tools are missing, the ToolSmith produces the configured tools like the weapons: First the priority orders and then the ratios (blue bars)

**Comment by codingberlin** [comment]
How I can prevent that the player can change the settings for other players, too?
Is there some code example for other partition based UI settings where the player is not able to change other players partition data?

**Comment by michaelzangl** [comment]
No. Such a test is currently not implemented in all of the partition-wise settings. Simply ignore it for now, it is on the TODO-List ;-)

**Comment by andreas-eberle** [comment]
@codingberlin: Maybe it would be an idea to add a check box like "activate automatic tool production" above the tool selection. With this check box, the user could decide, if he wants to use the automated tool production or not. Furthermore it would prevent conflicts like "user reduces tool amount, but it is increased again, as it is still needed".

**Comment by codingberlin** [comment]
@andreas-eberle Currently the automatic and manual tool production is merged in the code in the branch but there is a bug (I guess the bug is caused because the tool smith drops the tool at the beginning of the tool production animations and in the meantime it is rescheduled from the automation)

I would like to fix that by adding the tool to the production only once so that the automation adds tools but the human player can remove them if he likes (but why should he? maybe for hammers and shovels)

Then I would like to show you this and if you don't like it I can implement the checkbox.

**Comment by andreas-eberle** [comment]
@codingberlin: There is one issue with this idea: If you have a temporary high need of shovels or hammers (e.g. because you're just building a lot of stuff), then these shovels or hammers would be requested. However, at the time when they are actually produced, they aren't needed any more.

That's why the current system for the tool production calculates the currently needed tool in the moment the tool smith starts his work.

**Comment by michaelzangl** [comment]
All tool auto-production categories have this problem:
* Tools for building workers. They need to be requested as late as possible, because we often want to raze an old building instead of producing a new tool.
* Tools for building construction. Computing the optimal number of workers is not trivial.
* Weapons for recruiting (not implemented yet). No problem here, simply tie it to the number of weapons produced.

We might want to solve this in a more general way. The auto-production idea basically arose while we had no UI for it and simply needed enough materials. It is pretty advanced since it also selects the closest tool smith on big economies. This is a feature I do not want to miss and that we might want to allow the user to do it by himself (same with weapons).

My suggestion is to enhance the UI for more managed tool production (and weapon): When clicking on a building, you see the current list for that building (empty most of the time). You can schedule tools for the building to produce. They will be produced in that order.
Then we have the global production list. We add two "auto"-Buttons there, one for tools, one for weapons. For auto-construction-mode, we might need to use a absolute number of requested builders.
In auto mode, the up and down arrows disappear and the number only displays the materials that will be produced automatically. In manual mode, the user can change them.

That way, we have the ability to prevent automated production. We also have a nice queue system per house to allow expert players to control the game more.

**Comment by codingberlin** [comment]
I'm very open here how to implement the feature.
My target is to give the computerplayer an interface so that the WinnerGeneral can produce bows only (with some rush defending spears at the beginning)

**Comment by codingberlin** [comment]
As discussed with @andreas-eberle in icq I only implement the weapons here and the tools can be made within another issue/pullrequest.
